mne_bids/path.py

```diff
@@ -120,22 +132,88 @@ def __str__(self):
                     key = 'rec'
                 basename.append('%s-%s' % (key, val))
 
-        if self.suffix is not None:
-            basename.append(self.suffix)
+        if self.kind is not None:
+            if self.extension is not None:
+                basename.append(f'{self.kind}{self.extension}')
+            else:
+                basename.append(self.kind)
 
         basename = '_'.join(basename)
-        if self.prefix is not None:
-            basename = op.join(self.prefix, basename)
-
         return basename
 
+    @property
+    def fpath(self):
+        """Get the full filepath for this BIDS file.
+
+        Getting the file path consists of the following behavior:
+
+        1. ``bids_root`` is not passed into the object: A warning is shown
+        to the user and it return the ``basename``.
+
+        2. ``bids_root`` is passed into the object:
+
+            i. If ``kind`` and ``extension`` are given:
+            The function will then return the full filepath
+            using ``bids_root`` and ``basename``
+
+            ii. Else:
+            The function tries to infer the full filepath
+            for the given ``bids_root`` and ``basename``
+            and will return the full filepath. It will try to
+            infer the filename extension by searching
+            for the file on disk. If the file cannot be found, an error
+            will be raised. To disable this automatic inference attempt,
+            pass a string (like ``'.fif'`` or ``'.vhdr'``).
+
+            If it cannot find a matching dataset, it
+            will result in an error.
+
+        If ``None``,
+
+        Returns
+        -------
+        bids_fpath : str
+            Either the ``basename``, or full filepath to the dataset.
+        """
+        if self.bids_root is None:
+            msg = ('No bids root was passed in and '
+                   'full BIDS filepath cannot be '
+                   'automatically inferred. '
+                   'Returning the bids basename.')
+            warn(msg)
+            return self.basename
+
+        # TODO: how do we get rid of this? Hack to make sure no error out
+        # during write/read.
+        if self.kind is not None and self.extension is not None:
+            return op.join(self.bids_root, self.basename)
+
+        if self.kind is None:
+            msg = ('No kind was provided, and it cannot be '
+                   'automatically inferred. Please set kind to one of '
```

Comment:
maybe remove "and it cannot be automatically inferred" ?

in some cases, it can be right?

Reply:
If `kind` is not passed in, we wouldn't be able to infer the full file path. For example:

`* = sub-01_ses-01_task-01`:
```
*_ieeg.json
*_ieeg.edf
*_electrodes.tsv
*_coordsystem.json
*_channels.tsv
```

If `kind=None`, but everything else matched, do we return the `edf` file?

Reply:
Yeah since everything is "explicit", should we just axe this feature? Then if someone tries to pass a BIDsPath that really isn't a full valid file, then anything just errors out. We can always re-add. but rn I'm in favor of axing because it makes behavior very difficult to account for.

Reply:
Yes, back when I added it, I thought it was a good idea & simplification, but let's not try to be clever here and require users to be explicit. +1 for dropping this feature. Should this happen in another PR or would that lead to complete chaos here? I'm just concerned this PR is getting too big for us to manage!

Reply:
@adam2392 you have to return the favor by reviewing some PRs that @hoechenberger makes ;-)

Reply:
Sure just give me a request for review ;)